src/main.rs

```diff
@@ -716,363 +748,400 @@ fn main() -> Result<(), Box<dyn Error>> {
-    }
-
-    if !laser_defines_z {
+        let (mut electrons, mut photons, mut positrons) = primaries
+            .chunks(num / 20)
+            .enumerate()
+            .map(|(i, chk)| {
+                let tmp = chk.iter()
+                    .map(|pt| collide(&laser, *pt, &mut rng, &mut current_id, options))
+                    .fold((Vec::<Particle>::new(), Vec::<Particle>::new(), Vec::<Particle>::new()), merge);
+                if id == 0 {
+                    println!(
+                        "Done {: >12} of {: >12} primaries, RT = {}, ETTC = {}...",
+                        (i+1) * chk.len(), num,
+                        PrettyDuration::from(runtime.elapsed()),
+                        PrettyDuration::from(ettc(runtime, i+1, 20))
+                    );
+                }
+                tmp
+            })
+            .fold(
+                (Vec::<Particle>::new(), Vec::<Particle>::new(), Vec::<Particle>::new()),
+                |a, b| ([a.0,b.0].concat(), [a.1,b.1].concat(), [a.2,b.2].concat())
+            );
+
+        // Particle/parent ids are only unique within a single parallel process
+        let mut id_offsets = vec![0u64; world.size() as usize];
+        #[cfg(feature = "with-mpi")]
+        world.all_gather_into(&current_id, &mut id_offsets[..]);
+        id_offsets.iter_mut().fold(0, |mut total, n| {total += *n; *n = total - *n; total});
+        // task n adds id_offsets[n] to each particle/parent id
         for pt in electrons.iter_mut().chain(photons.iter_mut()).chain(positrons.iter_mut()) {
-            *pt = pt.to_beam_coordinate_basis(angle);
+            pt.with_id(pt.id() + id_offsets[id as usize]);
+            pt.with_parent_id(pt.parent_id() + id_offsets[id as usize]);
         }
-    }
 
-    for dstr in &eospec {
-        let prefix = format!("{}{}{}{}electron", output_dir, if output_dir.is_empty() {""} else {"/"}, ident, if ident.is_empty() {""} else {"_"});
-        dstr.write(&world, &electrons, &units, &prefix, file_format)?;
-    }
-
-    for dstr in &gospec {
-        let prefix = format!("{}{}{}{}photon", output_dir, if output_dir.is_empty() {""} else {"/"}, ident, if ident.is_empty() {""} else {"_"});
-        dstr.write(&world, &photons, &units, &prefix, file_format)?;
-    }
+        if !laser_defines_z {
+            for pt in electrons.iter_mut().chain(photons.iter_mut()).chain(positrons.iter_mut()) {
+                *pt = pt.to_beam_coordinate_basis(angle);
+            }
+        }
 
-    for dstr in &pospec {
-        let prefix = format!("{}{}{}{}positron", output_dir, if output_dir.is_empty() {""} else {"/"}, ident, if ident.is_empty() {""} else {"_"});
-        dstr.write(&world, &positrons, &units, &prefix, file_format)?;
-    }
+        for dstr in &eospec {
+            let prefix = format!("{}{}{}{}electron", output_dir, if output_dir.is_empty() {""} else {"/"}, ident, if ident.is_empty() {""} else {"_"});
```

Comment:
As `ident` is not changed while looping, this overwrites any distribution output that is produced.